File: cplusplus/src/safety_number.cpp
#include "safety_number.h"

#include "error.h"
#include "init.h"
#include "sizes.h"

namespace Autograph {

std::vector<unsigned char> calculateSafetyNumber(std::vector<unsigned char> a,
                                                 std::vector<unsigned char> b) {
  if (autograph_init() != 0) {
    throw Error(Error::Initialization);
  }
  std::vector<unsigned char> safetyNumber(SAFETY_NUMBER_SIZE);
  bool success =
      autograph_safety_number(safetyNumber.data(), a.data(), b.data()) == 0;
  if (!success) {
    throw Error(Error::SafetyNumberCalculation);
  }
  return safetyNumber;
}

}  // namespace Autograph

File: swift/Sources/Autograph/SafetyNumber.swift
import Clibautograph
import Foundation

public func calculateSafetyNumber(a: [UInt8], b: [UInt8]) throws -> [UInt8] {
  if autograph_init() < 0 {
    throw AutographError.initialization
  }
  var safetyNumber = createSafetyNumberBytes()
  let success =
    autograph_safety_number(
      &safetyNumber,
      a,
      b
    ) == 0
  if !success {
    throw AutographError.safetyNumberCalculation
  }
  return safetyNumber
}
